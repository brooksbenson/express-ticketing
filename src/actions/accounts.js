import db from '../firebase/firebase';

export const addAccount = ({ key, ...account }) => ({
  type: 'ADD_ACCOUNT',
  key,
  account
});

export const startAddAccount = account => {
  return async dispatch => {
    const { key } = await db.ref('account_data').push(account);
    dispatch(addAccount({ key, ...account }));
    return key;
  };
};

export const setAccounts = accounts => ({
  type: 'SET_ACCOUNTS',
  accounts
});

export const startSetAccounts = () => {
  return async dispatch => {
    const snap = await db.ref('account_data').once('value');
    dispatch(setAccounts(snap.val()));
  };
};

export const updateAccount = ({ key, ...update }) => ({
  type: 'UPDATE_ACCOUNT',
  key,
  update
});

export const startUpdateAccount = ({ key, ...update }) => {
  return async dispatch => {
    await db.ref(`account_data/${key}`).set(update);
    dispatch(updateAccount({ key, ...update }));
  };
};
