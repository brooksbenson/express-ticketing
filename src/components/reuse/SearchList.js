import React from 'react';

export default class SearchList extends React.Component {
  state = { search: '' };

  onSearch = e => {
    const search = e.target.value;
    this.setState(() => ({ search }));
  };

  render() {
    const { list, selector, className } = this.props;
    const { search } = this.state;
    return (
      <div className={className}>
        <input
          className="search"
          onChange={this.onSearch}
          placeholder="Search..."
          type="text"
          value={search}
        />
        <ul>
          {selector(list, search).map(item => (
            <li key={a.key} onClick={() => this.props.onClick(item)}>
              <button>{a.name}</button>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}
