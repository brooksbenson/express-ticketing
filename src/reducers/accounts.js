export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ACCOUNT':
      return [...state, action.account];
    case 'SET_ACCOUNTS':
      return action.accounts;
    default:
      return state;
  }
};
