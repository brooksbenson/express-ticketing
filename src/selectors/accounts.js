export default (accounts, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return accounts.filter(a => regex.test(a.name));
};
