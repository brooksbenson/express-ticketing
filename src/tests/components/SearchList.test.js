import React from 'react';
import { shallow } from 'enzyme';
import SearchList from '../../components/reuse/SearchList';
import accounts from '../fixtures/accounts';
import accountSelector from '../../selectors/accounts';

let wrapper;
let onClick;
beforeEach(() => {
  onClick = jest.fn();
  wrapper = shallow(
    <SearchList
      className="test__list"
      list={accounts}
      onClick={onClick}
      selector={accountSelector}
    />
  );
});

test('should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('wrapper should have class "test__list"', () => {
  expect(wrapper.shouldHaveClass('test__list')).toBeTruthy();
});

test('state should change on search', () => {
  const value = 'foo';
  wrapper.find('input').simulate('change', { target: { value } });
  expect(wrapper.state('search')).toBe(value);
});

test('list length should differ on search', () => {
  const value = 'jfdksa;fdka;fj';
  const preLength = wrapper.find('ul').children().length;
  wrapper.find('input').simulate('change', { target: { value } });
  const postLength = wrapper.find('ul').children().length;
  expect(postLength).toBeLessThan(preLength);
});

test('onClick should fire on list item click', () => {
  wrapper
    .find('li')
    .at(0)
    .simulate('click');
  expect(onClick).toHaveBeenCalled();
});
