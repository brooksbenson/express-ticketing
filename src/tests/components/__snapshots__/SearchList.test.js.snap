// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="test__list"
>
  <input
    className="search"
    onChange={[Function]}
    placeholder="Search..."
    type="text"
    value=""
  />
  <ul>
    <li
      key="jfkda;"
      onClick={[Function]}
    >
      <button>
        Flower Motors
      </button>
    </li>
    <li
      key="kfjdka"
      onClick={[Function]}
    >
      <button>
        Prophecy Systems
      </button>
    </li>
    <li
      key="fjdkal"
      onClick={[Function]}
    >
      <button>
        Dynamic Foods
      </button>
    </li>
  </ul>
</div>
`;
